python/ray/data/datasource/mongo_datasource.py

```diff
@@ -0,0 +1,185 @@
+from ray.data.block import (
+    Block,
+    BlockMetadata,
+)
+from ray.data._internal.remote_fn import cached_remote_fn
+from ray.types import ObjectRef
+from ray.util.annotations import PublicAPI
+
+import pymongo
+from pymongoarrow.api import Schema
+
+logger = logging.getLogger(__name__)
+
+
+@PublicAPI(stability="alpha")
+class MongoDatasource(Datasource):
+    """Datasource for reading from and writing to MongoDB.
+
+    Examples:
+        >>> import ray
+        >>> from ray.data.datasource import MongoDatasource
+        >>> from pymongoarrow.api import Schema
+        >>> ds = ray.data.read_datasource( # doctest: +SKIP
+        ...     MongoDatasource(),
+        ...     uri="mongodb://username:password@mongodb0.example.com:27017/?authSource=admin", # noqa: E501,
+        ...     database="my_db",
+        ...     collection="my_collection",
+        ...     schema=Schema({"col1": pa.string(), "col2": pa.int64()}),
+        ... )
+    """
+
+    def create_reader(self, **kwargs) -> Reader:
+        return _MongoDatasourceReader(**kwargs)
+
+    def do_write(
+        self,
+        blocks: List[ObjectRef[Block]],
+        metadata: List[BlockMetadata],
+        ray_remote_args: Optional[Dict[str, Any]],
+        uri: str,
+        database: str,
+        collection: str,
+    ) -> List[ObjectRef[WriteResult]]:
+        _validate_database_collection_exist(
+            pymongo.MongoClient(uri), database, collection
+        )
+
+        def write_block(uri: str, database: str, collection: str, block: Block):
+            from pymongoarrow.api import write
+
+            client = pymongo.MongoClient(uri)
+            write(client[database][collection], block)
+
+        if ray_remote_args is None:
+            ray_remote_args = {}
+
+        write_block = cached_remote_fn(write_block).options(**ray_remote_args)
+        write_tasks = []
+        for block in blocks:
+            write_task = write_block.remote(uri, database, collection, block)
+            write_tasks.append(write_task)
+        return write_tasks
+
+
+class _MongoDatasourceReader(Reader):
+    def __init__(
+        self,
+        uri: str,
+        database: str,
+        collection: str,
+        pipeline: List[Dict] = None,
+        schema: Schema = None,
+        **mongo_args,
+    ):
+        self._uri = uri
+        self._database = database
+        self._collection = collection
+        self._pipeline = pipeline
+        self._schema = schema
+        self._mongo_args = mongo_args
+        # If pipeline is unspecified, read the entire collection.
+        if not pipeline:
+            self._pipeline = [{"$match": {"_id": {"$exists": "true"}}}]
+
+        self._client = pymongo.MongoClient(uri)
+        _validate_database_collection_exist(self._client, database, collection)
+
+    def estimate_inmemory_data_size(self) -> Optional[int]:
+        # TODO(jian): Add memory size estimation to improve auto-tune of parallelism.
```

Comment:
https://www.mongodb.com/docs/manual/reference/command/collStats/#collstats can give statistics in the collection, that might be helpful.